There is an implicit cast in the code:

```
/builds/hzdr/crp/alpaka/test/unit/math/src/powMixedTypes.cpp:47:39: error: implicit conversion increases floating-point precision: 'float' to 'const double' [-Werror,-Wdouble-promotion]
        return static_cast<TExpected>(0.0f);
               ~~~~~~~~~~~            ^~~~
/builds/hzdr/crp/alpaka/test/unit/math/src/powMixedTypes.cpp:29:66: note: in instantiation of function template specialization 'PowMixedTypesTestKernel<alpaka::Complex<double>>::toExpected<alpaka::Complex<float>>' requested here
        auto expected = alpaka::math::pow(acc, toExpected(arg1), toExpected(arg2));
                                                                 ^
/usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/invoke.h:60:14: note: in instantiation of function template specialization 'PowMixedTypesTestKernel<alpaka::Complex<double>>::operator()<alpaka::AccCpuSerial<std::integral_constant<unsigned long, 1>, unsigned long>, double, alpaka::Complex<float>>' requested here
    { return std::forward<_Fn>(__f)(std::forward<_Args>(__args)...); }
```

--- a/include/alpaka/math/MathUniformCudaHipBuiltIn.hpp
+++ b/include/alpaka/math/MathUniformCudaHipBuiltIn.hpp
@@ -721,7 +721,7 @@ namespace alpaka::math
                 if constexpr(is_decayed_v<TBase, float> && is_decayed_v<TExp, float>)
                     return ::powf(base, exp);
                 else if constexpr(is_decayed_v<TBase, double> || is_decayed_v<TExp, double>)
-                    return ::pow(base, exp);
+                    return ::pow(static_cast<double>(base), static_cast<double>(exp));
                 else
                     static_assert(!sizeof(TBase), "Unsupported data type");
 

--- a/test/unit/math/src/Defines.hpp
+++ b/test/unit/math/src/Defines.hpp
@@ -98,6 +98,26 @@ namespace alpaka
                     return arg1 + arg2;
                 }
 
+                // https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon
+                template<typename TAcc, typename FP>
+                ALPAKA_FN_ACC auto almost_equal(TAcc const& acc, FP x, FP y, int ulp)
+                    -> std::enable_if_t<!std::numeric_limits<FP>::is_integer, bool>
+                {
+                    // the machine epsilon has to be scaled to the magnitude of the values used
+                    // and multiplied by the desired precision in ULPs (units in the last place)
+                    return alpaka::math::abs(acc, x - y)
+                        <= std::numeric_limits<FP>::epsilon() * alpaka::math::abs(acc, x + y) * static_cast<FP>(ulp)
+                        // unless the result is subnormal
+                        || alpaka::math::abs(acc, x - y) < std::numeric_limits<FP>::min();
+                }
+
+                //! Version for alpaka::Complex
+                template<typename TAcc, typename FP>
+                ALPAKA_FN_ACC bool almost_equal(TAcc const& acc, alpaka::Complex<FP> x, alpaka::Complex<FP> y, int ulp)
+                {
+                    return almost_equal(acc, x.real(), y.real(), ulp) && almost_equal(acc, x.imag(), y.imag(), ulp);
+                }
+
             } // namespace math
         } // namespace unit
     } // namespace test

--- a/test/unit/math/src/powMixedTypes.cpp
+++ b/test/unit/math/src/powMixedTypes.cpp
@@ -7,6 +7,8 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
 
+#include "Defines.hpp"
+
 #include <alpaka/math/Traits.hpp>
 #include <alpaka/test/KernelExecutionFixture.hpp>
 #include <alpaka/test/acc/TestAccs.hpp>
@@ -15,26 +17,6 @@
 #include <catch2/catch.hpp>
 
 #include <type_traits>
-
-// https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon
-template<typename TAcc, typename FP>
-ALPAKA_FN_ACC auto almost_equal(TAcc const& acc, FP x, FP y, int ulp)
-    -> std::enable_if_t<!std::numeric_limits<FP>::is_integer, bool>
-{
-    // the machine epsilon has to be scaled to the magnitude of the values used
-    // and multiplied by the desired precision in ULPs (units in the last place)
-    return alpaka::math::abs(acc, x - y)
-        <= std::numeric_limits<FP>::epsilon() * alpaka::math::abs(acc, x + y) * static_cast<FP>(ulp)
-        // unless the result is subnormal
-        || alpaka::math::abs(acc, x - y) < std::numeric_limits<FP>::min();
-}
-
-//! Version for alpaka::Complex
-template<typename TAcc, typename FP>
-ALPAKA_FN_ACC bool almost_equal(TAcc const& acc, alpaka::Complex<FP> x, alpaka::Complex<FP> y, int ulp)
-{
-    return almost_equal(acc, x.real(), y.real(), ulp) && almost_equal(acc, x.imag(), y.imag(), ulp);
-}
 
 template<typename TExpected>
 class PowMixedTypesTestKernel
@@ -46,6 +28,7 @@ public:
     {
         auto expected = alpaka::math::pow(acc, TExpected{arg1}, TExpected{arg2});
         auto actual = alpaka::math::pow(acc, arg1, arg2);
+        using alpaka::test::unit::math::almost_equal;
         ALPAKA_CHECK(*success, almost_equal(acc, expected, actual, 1));
     }
 };

--- a/test/unit/math/src/powMixedTypes.cpp
+++ b/test/unit/math/src/powMixedTypes.cpp
@@ -18,6 +18,39 @@
 
 #include <type_traits>
 
+//! Convert the given real or complex input type to the given real or complex output type avoiding warnings.
+//! This conversion is surprisingly tricky to do in a way that no compiler complains.
+//! In principle it could be accomplished by a constexpr function, but in practice that turned out not possible.
+//! The general implementation does direct initialization, works for pairs of types supporting it.
+template<typename TInput, typename TOutput, typename TSfinae = void>
+struct Convert
+{
+    ALPAKA_FN_ACC auto operator()(TInput const arg) const
+    {
+        return TOutput{arg};
+    }
+};
+
+//! Specialization for real -> real conversion
+template<typename TInput, typename TOutput>
+struct Convert<TInput, TOutput, std::enable_if_t<std::is_floating_point_v<TOutput>>>
+{
+    ALPAKA_FN_ACC auto operator()(TInput const arg) const
+    {
+        return static_cast<TOutput>(arg);
+    }
+};
+
+//! Specialization for real -> complex conversion
+template<typename TInput, typename TOutputValueType>
+struct Convert<TInput, alpaka::Complex<TOutputValueType>, std::enable_if_t<std::is_floating_point_v<TInput>>>
+{
+    ALPAKA_FN_ACC auto operator()(TInput const arg) const
+    {
+        return alpaka::Complex<TOutputValueType>{static_cast<TOutputValueType>(arg)};
+    }
+};
+
 template<typename TExpected>
 class PowMixedTypesTestKernel
 {
@@ -26,7 +59,7 @@ public:
     template<typename TAcc, typename TArg1, typename TArg2>
     ALPAKA_FN_ACC auto operator()(TAcc const& acc, bool* success, TArg1 const arg1, TArg2 const arg2) const -> void
     {
-        auto expected = alpaka::math::pow(acc, TExpected{arg1}, TExpected{arg2});
+        auto expected = alpaka::math::pow(acc, Convert<TArg1, TExpected>{}(arg1), Convert<TArg2, TExpected>{}(arg2));
         auto actual = alpaka::math::pow(acc, arg1, arg2);
         using alpaka::test::unit::math::almost_equal;
         ALPAKA_CHECK(*success, almost_equal(acc, expected, actual, 1));

--- a/test/unit/math/src/sincos.cpp
+++ b/test/unit/math/src/sincos.cpp
@@ -7,6 +7,8 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
 
+#include "Defines.hpp"
+
 #include <alpaka/math/Traits.hpp>
 #include <alpaka/test/KernelExecutionFixture.hpp>
 #include <alpaka/test/acc/TestAccs.hpp>
@@ -16,25 +18,6 @@
 
 #include <type_traits>
 
-// https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon
-template<typename TAcc, typename FP>
-ALPAKA_FN_ACC auto almost_equal(TAcc const& acc, FP x, FP y, int ulp)
-    -> std::enable_if_t<!std::numeric_limits<FP>::is_integer, bool>
-{
-    // the machine epsilon has to be scaled to the magnitude of the values used
-    // and multiplied by the desired precision in ULPs (units in the last place)
-    return alpaka::math::abs(acc, x - y)
-        <= std::numeric_limits<FP>::epsilon() * alpaka::math::abs(acc, x + y) * static_cast<FP>(ulp)
-        // unless the result is subnormal
-        || alpaka::math::abs(acc, x - y) < std::numeric_limits<FP>::min();
-}
-
-//! Version for alpaka::Complex
-template<typename TAcc, typename FP>
-ALPAKA_FN_ACC bool almost_equal(TAcc const& acc, alpaka::Complex<FP> x, alpaka::Complex<FP> y, int ulp)
-{
-    return almost_equal(acc, x.real(), y.real(), ulp) && almost_equal(acc, x.imag(), y.imag(), ulp);
-}
 
 class SinCosTestKernel
 {
@@ -50,6 +33,7 @@ public:
         FP result_sin = 0.;
         FP result_cos = 0.;
         alpaka::math::sincos(acc, arg, result_sin, result_cos);
+        using alpaka::test::unit::math::almost_equal;
         ALPAKA_CHECK(
             *success,
             almost_equal(acc, result_sin, check_sin, 1) && almost_equal(acc, result_cos, check_cos, 1));

